### ISSUE TITLE
node-debug very delayed source file hydration

### ISSUE THREAD

**Issue by samccone** [node-inspector/node-inspector#498]
I am running node-inspector 0.8.1

The command I am running is
`node-debug --hidden='[\"node_modules\"]' -–no-preload --debug-brk=0 `

After around 2 min the sources tab fills in with my files.

Even though I specify node_modules as hidden the node_modules are still showing in the sources.

![screen shot 2014-12-17 at 4 54 52 pm](https://cloud.githubusercontent.com/assets/883126/5479683/8741144a-860d-11e4-9450-bb02844fa464.png)

I suspect this is the real cause of the slow boot up time. Any insight would be greatly appreciated, and thanks so much for this great tool!

**Comment by bajtos** [comment]
It seems like `node-debug` is not correctly forwarding `--hidden` option to Node Inspector.

**Comment by samccone** [comment]
awesome! Let me know if I can help out at all in debugging

**Comment by samccone** [comment]
Alright after using the work done in #530 it seems like this problem is still around

![screen shot 2015-01-22 at 10 33 08 am](https://cloud.githubusercontent.com/assets/883126/5858566/2c773ea0-a222-11e4-88c5-beb237c0b6b5.png)

As you can see it takes around 2 min for the UI with my files to load!

After looking around a bit... this line seems highly suspicious 

https://github.com/node-inspector/node-inspector/blame/813354c66e1fee70012aae94863ead0d3eb68abe/front-end/utilities.js#L1297

thoughts?

**Comment by bajtos** [comment]
My take is that the line 1297 is where the front-end is loading the content of all source files in your project. I guess Chrome developers assume that the files (scripts) are already loaded in the browser, thus its cheap and fast to get them from the backend. Unfortunately that's not the case of Node Inspector, which has to load all of them from the filesystem.

I see two ways how to work around that:
 - reduce the number of files served to front-end (e.g. via `--hidden="node_modules"`)
 - contribute a change to Chromium DevTools front-end and make it load files lazily, only after the user asked front-end to open the file

**Comment by jldec** [comment]
I had a similar problem (2-5 minute delays on debugger startup) and did a little digging.

The issue in my case came from giant glob results (cost ~40s) and subseqent filtering (cost ~2 min!) on the backend.

A quick fix for me was to update to the latest version of glob (which stops symlink exploration - my project has quite a few local npm packages linked via `npm link`). But since i do need to debug to one level of symlinks i changed the glob patterns from `**/*`to `**/**/*` in ScriptFileStorage.js which exposes what i needed for debugging. 

    var CONVENTIONAL_DIRS_PATTERN = '{*.js,lib/**/*.js,node_modules/**/**/*.js,test/**/*.js}';
    var ALL_JS_PATTERN = '**/**/*.js';

Now i get everything i need in about 5s... much better. Not sure it's a universal fix, but h.t.h.

refs:
- https://github.com/isaacs/node-glob/issues/134
- https://github.com/ericf/express-handlebars/pull/98
- https://github.com/adgad/express-handlebars/commit/803129ed9b325640cc527aefdb5096dda7661355

PS - There are definitely ways to speed up the filtering as well in `findAllApplicationScripts()`

**Comment by bajtos** [comment]
One of the root causes is inefficient filtering algorithm in [ScriptFileStorage.findAllApplicationScripts()](https://github.com/node-inspector/node-inspector/blob/7d1f207e922a6bf427d25066bf956c4cb6d247cd/lib/ScriptFileStorage.js#L229-L231):

```js
      files = files.filter(function(elem, ix, arr) {
        return arr.indexOf(elem) >= ix && !this._scriptManager.isScriptHidden(elem);
      }.bind(this));
```

Combination of `files.filter` and `arr.indexOf(elem)` has speed complexity of `O(n^2)`, which scales poorly and explains why the filtering can take several minutes.

**Comment by yoshiokatsuneo** [comment]
For me, adding preload and hidden helped.

(Gruntfile.js)
```
    'node-inspector': {
      custom: {
        options: {
          'web-host': 'localhost',
          preload: false,
          hidden: ["node_modules/"],
        }
      }
    },
```

**Comment by 3y3** [comment]
Should be fixed in 0.12.6

**Comment by samccone** [comment]
awesome thanks @3y3 !
